internal/ecies_ethereum.go

```diff
@@ -91,30 +91,9 @@ func (prv *PrivateKeyPkcs11) Public() *ecies.PublicKey {
 }
 
 // NIST SP 800-56 Concatenation Key Derivation Function (see section 5.8.1).
-func concatKDF(hash hash.Hash, z, s1 []byte, kdLen int) []byte {
-	counterBytes := make([]byte, 4)
-	k := make([]byte, 0, roundup(kdLen, hash.Size()))
-	for counter := uint32(1); len(k) < kdLen; counter++ {
-		binary.BigEndian.PutUint32(counterBytes, counter)
-		hash.Reset()
-		if _, err := hash.Write(counterBytes); err != nil {
-			return nil
-		}
-		if _, err := hash.Write(z); err != nil {
-			return nil
-		}
-		if _, err := hash.Write(s1); err != nil {
-			return nil
-		}
-		k = hash.Sum(k)
-	}
-	return k[:kdLen]
-}
-
-// roundup rounds size up to the next multiple of blocksize.
-func roundup(size, blocksize int) int {
-	return size + blocksize - (size % blocksize)
-}
+//
+//go:linkname concatKDF github.com/umbracle/ecies.concatKDF
+func concatKDF(hash hash.Hash, z, s1 []byte, kdLen int) []byte
```

Comment:
looks like a hack :).

Reply:
Yes, it is... still the Golang standard library uses it extensively. So we should treat it as safe from any perspective.
Every time I tried to do something wrong on this line - either the compiler blamed on me, or the code panicked. Both ways to fail seem good to me.

Reply:
Anyway, I consider this temporary if we go for forking `umbracle/ecies` and injecting our pkcs11 changes into it.